src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,583 @@
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Utils;
+// ReSharper disable AccessToDisposedClosure
+
+namespace Raven.Server.Documents.ShardedTcpHandlers
+{
+    public class ShardedSubscriptionConnection : SubscriptionConnectionBase, IDisposable
+    {
+        private List<SubscriptionServerWorker> _workers;
+        private bool _isDisposed;
+
+        private ShardedSubscriptionConnection(ServerStore serverStore, TcpConnectionOptions tcpConnectionOptions, IDisposable tcpConnectionDisposable, JsonOperationContext.MemoryBuffer buffer)
+            : base(tcpConnectionOptions, serverStore, buffer, tcpConnectionDisposable)
+        {
+        }
+
+        public static void SendShardedSubscriptionDocuments(ServerStore server, TcpConnectionOptions tcpConnectionOptions, JsonOperationContext.MemoryBuffer buffer)
+        {
+            var remoteEndPoint = tcpConnectionOptions.TcpClient.Client.RemoteEndPoint;
+
+            var tcpConnectionDisposable = tcpConnectionOptions.ConnectionProcessingInProgress("ShardedSubscription");
+            try
+            {
+                var connection = new ShardedSubscriptionConnection(server, tcpConnectionOptions, tcpConnectionDisposable, buffer);
+                try
+                {
+                    Task.Run(async () =>
+                    {
+                        using (tcpConnectionOptions)
+                        using (tcpConnectionDisposable)
+                        using (connection)
+                        {
+                            connection._lastConnectionStats = new SubscriptionConnectionStatsAggregator(_connectionStatsId, null);
+                            connection._connectionScope = connection._lastConnectionStats.CreateScope();
+                            connection._pendingConnectionScope = connection._connectionScope.For(SubscriptionOperationScope.ConnectionPending);
+                            try
+                            {
+                                try
+                                {
+                                    await connection.InitAsync();
+                                    await connection.ProcessSubscriptionAsync();
+                                }
+                                catch (SubscriptionInvalidStateException)
+                                {
+                                      connection._pendingConnectionScope.Dispose();
+                                    throw;
+                                }
+
+                            }
+                            catch (Exception e)
+                            {
+                                connection._connectionScope.RecordException(e is SubscriptionInUseException ? SubscriptionError.ConnectionRejected : SubscriptionError.Error, e.Message);
+
+                                var errorMessage = $"Failed to process sharded subscription {connection.SubscriptionId} / from client {remoteEndPoint}";
```

Comment:
`remoteEndPoint` is bad idea, we know what the actual URL is and what node this is, let's use this.

Reply:
still an issue